jiant/tasks/lib/templates/span_prediction.py

```diff
@@ -0,0 +1,172 @@
+    token_idx_to_char_idx_map: np.ndarray
+
+    def featurize(self, tokenizer, feat_spec):
+
+        if feat_spec.sep_token_extra:
+            maybe_extra_sep = [tokenizer.sep_token]
+            maybe_extra_sep_segment_id = [feat_spec.sequence_a_segment_id]
+            special_tokens_count = 4  # CLS, SEP-SEP, SEP
+        else:
+            maybe_extra_sep = []
+            maybe_extra_sep_segment_id = []
+            special_tokens_count = 3  # CLS, SEP, SEP
+
+        passage, question = truncate_sequences(
+            tokens_ls=[self.passage, self.question],
+            max_length=feat_spec.max_seq_length - special_tokens_count,
+        )
+        assert (
+            len(passage) >= self.answer_token_span[1]
+        ), f"Answer span {self.answer_token_span} truncated, please raise max_seq_length."
+        unpadded_inputs = add_cls_token(
+            unpadded_tokens=(
+                passage + [tokenizer.sep_token] + maybe_extra_sep + question + [tokenizer.sep_token]
+            ),
+            unpadded_segment_ids=(
+                [feat_spec.sequence_a_segment_id] * (len(passage) + 1)
+                + maybe_extra_sep_segment_id
+                + [feat_spec.sequence_b_segment_id] * (len(question) + 1)
+            ),
+            tokenizer=tokenizer,
+            feat_spec=feat_spec,
+        )
+        gt_span_idxs = list(map(lambda x: x + unpadded_inputs.cls_offset, self.answer_token_span))
+        input_set = create_input_set_from_tokens_and_segments(
+            unpadded_tokens=unpadded_inputs.unpadded_tokens,
+            unpadded_segment_ids=unpadded_inputs.unpadded_segment_ids,
+            tokenizer=tokenizer,
+            feat_spec=feat_spec,
+        )
+        pred_span_mask = pad_to_max_seq_length(
+            ls=[0] * unpadded_inputs.cls_offset + [1] * len(passage),
+            max_seq_length=feat_spec.max_seq_length,
+            pad_idx=0,
+            pad_right=not feat_spec.pad_on_left,
+        )
+        ls = [-1] * unpadded_inputs.cls_offset + (self.token_idx_to_char_idx_map > 0).argmax(
+            axis=1
+        ).tolist()[: len(passage)]
+        token_idx_to_char_idx_start = pad_to_max_seq_length(
+            ls=ls,
+            max_seq_length=feat_spec.max_seq_length,
+            pad_idx=-1,
+            pad_right=not feat_spec.pad_on_left,
+        )
+        token_idx_to_char_idx_end = pad_to_max_seq_length(
+            ls=[-1] * unpadded_inputs.cls_offset
+            + self.token_idx_to_char_idx_map.cumsum(axis=1).argmax(axis=1).tolist()[: len(passage)],
+            max_seq_length=feat_spec.max_seq_length,
+            pad_idx=-1,
+            pad_right=not feat_spec.pad_on_left,
+        )
+        # when there are multiple greatest elements, argmax will return the index of the first one
+        # so, (x > 0).argmax() will return the index of the first non-zero element in an array
+        # and x.cumsum().argmax() will return the index of the last non-zero element in an array
+
```

Comment:
A comment here would be helpful to explain what `token_idx_to_char_idx_start` and `token_idx_to_char_idx_end` represent (I found it helpful to print `list(zip(token_idx_to_char_idx_start, token_idx_to_char_idx_end))` to see that together these represent the character spans in the original passage text for each of the tokens (including padding)).